### PR TITLE
feat: ArgoCD HA chart reusing existing charts

### PR DESCRIPTION
**Pull request by carlosjgp**
Would say that this is far from ideal but it renders a set of resources that are equivalent to the [manifests provided on the argo-cd repository](https://github.com/argoproj/argo-cd/tree/v1.4.2/manifests/ha)

Fixes (ish) #172 

I would like to have some feedback about this approach that is not as neat and nice as the [manifest rendered with Kustomize](https://github.com/argoproj/argo-cd/tree/v1.4.2/manifests/ha)

The other approach could be to create the Helm chart from scratch or maybe using Kustomize to implement the shared bits for HA and non HA...

Thoughts?

Signed-off-by: Carlos Juan Gómez Peñalver <carlosjuangp@gmail.com>

Checklist:

* [x] I have update the chart version in `Chart.yaml` following Semantic Versioning.
* [x] Any new values are backwards compatible and/or have sensible default.
* [x] I have followed the testing instructions in the [contributing guide](https://github.com/argoproj/argo-helm/blob/master/CONTRIBUTING.md).
* [x] I have signed the CLA and the build is green.
* [x] I will test my changes again once merged to master and published.

Changes are automatically published when merged to `master`. They are not published on branches.

### CODE REVIEW

**Comment by cabrinha** [comment]
I'm not sure if it'll make much sense seeing as how we can reference the argo-cd chart, but adding a README in here would be nice: https://github.com/norwoodj/helm-docs

**Comment by seanson** [comment]
I think I'd rather have this configured in the ArgoCD via flags as maintaining two charts would be difficult.

**Comment by carlosjgp** [comment]
> I think I'd rather have this configured in the ArgoCD chart via flags as maintaining two charts would be difficult.

Sounds good.

I could use the Redis HA stable on the ArgoCD chart. I'll update this PR to do so.

Will this change be ok for the Argo team?

File: charts/argo-cd-ha/values.yaml [hunk 1]
```diff
@@ -0,0 +1,53 @@
+argo-cd:
+  installCRDs: false
+  controller:
+    extraArgs:
+      sentinel=argocd-redis-ha-announce-0:
```
Comment (cabrinha) [inline charts/argo-cd-ha/values.yaml]: Maybe we should handle the key/value mapping differently if the redis-ha chart is added to the argo-cd chart... Something like if redis-ha.enabled -- We could be more clear about adding these args into the deployments.

**Comment by cabrinha** [comment]
I tried out this setup and ArgoCD seems to be having a problem managing itself and rendering certain parts of this chart:

1. Sentinel configuration via `extraArgs`
Screen Shot 2020-02-19 at 8 11 21 AM

When I render the chart locally via `helm template . -f values.yaml` these sentinel flags render fine. But, when this chart is applied via ArgoCD, these flags fail to render.

This remains true for all three deployment manifests that use redis sentinel. The application-controller, the repo-server and the server.

2. HPA fails to reach a "Synced" state

```
level=warning msg="Failed to convert resource: no kind \"APIService\" is registered for version \"apiregistration.k8s.io/v1\" in scheme \"k8s.io/kubernetes/pkg/api/legacyscheme/scheme.go:30\""
```

Screen Shot 2020-02-19 at 9 30 53 AM

For some reason, ArgoCD feels the need to re-order the HPA manifest to have the memory come before the cpu spec.

**Comment by carlosjgp** [comment]
> I tried out this setup and ArgoCD seems to be having a problem managing itself and rendering certain parts of this chart:

The hacky expression for `extraArgs` is due to this piece of code
https://github.com/argoproj/argo-helm/blob/master/charts/argo-cd/templates/argocd-application-controller/deployment.yaml#L59-L65
and the fact that this field is using a `map[string]string` instead of just `[]string` which means that you can't use repeated keys (`sentinel`) will be resolved into only one, the last one.
I might raise a ticket for this _breaking change_ separately and then correct this PR.
I haven't checked ArgoCD code but depending on how is marshalling and unmarshalling YAML what is valid for `helm template ...` might not be for ArgoCD
(I'll check this and come back with a more concrete answer for these :point_up: )

Could you provide your K8s cluster details, please?

```
$ kubectl version 
```

> level=warning msg="Failed to convert resource: no kind \"APIService\" is registered for version \"apiregistration.k8s.io/v1\" in scheme \"k8s.io/kubernetes/pkg/api/legacyscheme/scheme.go:30\""

Are you running `metrics-server` ? This message and problems with Horizontal Pod Autoscaler are normally related to old `metrics-server` versions or even using `heaspter` (deprecated now)

**Comment by cabrinha** [comment]
> > I tried out this setup and ArgoCD seems to be having a problem managing itself and rendering certain parts of this chart:
> 
> The hacky expression for `extraArgs` is due to this piece of code
> https://github.com/argoproj/argo-helm/blob/master/charts/argo-cd/templates/argocd-application-controller/deployment.yaml#L59-L65
> and the fact that this field is using a `map[string]string` instead of just `[]string` which means that you can't use repeated keys (`sentinel`) will be resolved into only one, the last one.
> I might raise a ticket for this _breaking change_ separately and then correct this PR.
> I haven't checked ArgoCD code but depending on how is marshalling and unmarshalling YAML what is valid for `helm template ...` might not be for ArgoCD
> (I'll check this and come back with a more concrete answer for these ☝️ )
> 
> Could you provide your K8s cluster details, please?
> 
> ```
> $ kubectl version 
> ```
> 
> > level=warning msg="Failed to convert resource: no kind "APIService" is registered for version "apiregistration.k8s.io/v1" in scheme "k8s.io/kubernetes/pkg/api/legacyscheme/scheme.go:30""
> 
> Are you running `metrics-server` ? This message and problems with Horizontal Pod Autoscaler are normally related to old `metrics-server` versions or even using `heaspter` (deprecated now)

```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"06ad960bfd03b39c8310aaf92d1e7c12ce618213", GitTreeState:"clean", BuildDate:"2020-02-13T18:08:14Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.9-eks-c0eccc", GitCommit:"c0eccca51d7500bb03b2f163dd8d534ffeb2f7a2", GitTreeState:"clean", BuildDate:"2019-12-22T23:14:11Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
```

File: charts/argo-cd-ha/values.yaml (hunk 1 above)
Comment (carlosjgp) [reply charts/argo-cd-ha/values.yaml]: The "problem" with that is to find the balance between how much you want to hide behind that flag and how much flexibility you give to the user to deploy the chart...
But yes, it's a good idea. I'll try some GoLang templates (aka `_helpers.tpl`) to see if I can get that flexibility and default values

Comment (carlosjgp) [reply charts/argo-cd-ha/values.yaml]: I raised a PR to improve this weird syntax

https://github.com/argoproj/argo-helm/pull/244

**Comment by cabrinha** [comment]
> > I think I'd rather have this configured in the ArgoCD chart via flags as maintaining two charts would be difficult.
> 
> Sounds good.
> 
> I could use the Redis HA stable on the ArgoCD chart. I'll update this PR to do so.
> 
> Will this change be ok for the Argo team?

On the releases page for ArgoCD, there are two options for installation: HA and Non-HA.

I think this helm chart should leave these two options open to the user, with the [redis-ha](https://github.com/helm/charts/tree/master/stable/redis-ha) chart being enabled for the HA option and the [redis](https://github.com/helm/charts/tree/master/stable/redis) chart being enabled for the Non-HA option.

**Comment by cabrinha** [comment]
This PR brings up some important points: https://github.com/argoproj/argo-cd/pull/3147

**Comment by cabrinha** [comment]
@carlosjgp -- Now that https://github.com/argoproj/argo-cd/pull/3147 has been merged, it's probably time to get a major update to the argo-cd chart that mirrors that setup.

**Comment by carlosjgp** [comment]
Superseded by #270

**Pull request closed without merging**